File: baekjoon/python/gold3/q1005_acm_craft.py
# https://www.acmicpc.net/problem/1005
# 5/33
from collections import deque
import copy
import sys
input = sys.stdin.readline


def solution():
    for _ in range(int(input())):
        n, k = map(int, input().split())
        t = [0]
        for i in map(int, input().split()) :
            t.append(i)
        indegree = [0] * (n+1)

        board = [[] for _ in range(n+1)]
        for _ in range(k):
            x, y = map(int, input().split())
            board[x].append(y)
            indegree[y] += 1
        
        w = int(input())
        result = copy.deepcopy(t)
        q = deque()
        for i in range(1, n+1):
            if indegree[i] == 0:
                q.append(i)
        
        while q:
            now = q.popleft()
            for i in board[now]:
                result[i] = max(result[i], result[now] + t[i])
                indegree[i] -= 1
                if indegree[i] == 0 :
                    q.append(i)

        print(result[w])


sys.setrecursionlimit(10000)
def solution_time_fail():
    global t
    global answer
    global board

    for _ in range(int(input())):
        answer = []
        n, k = map(int, input().split())
        t = [0]
        for i in map(int, input().split()) :
            t.append(i)

        board = [[] for _ in range(n+1)]
        for _ in range(k):
            x, y = map(int, input().split())
            board[y].append(x)

        w = int(input())
        dfs(w, 0)
        answer.sort(reverse=True)
        print(answer[0])

def dfs(now, time):
    global answer
    global board
    time += t[now]
    if board[now] == []:
        answer.append(time)
        return
    
    for to in board[now]:
        dfs(to, time)

def solution_memory_fail():
    for _ in range(int(input())):
        n, k = map(int, input().split())
        t = [0]
        for i in map(int, input().split()) :
            t.append(i)

        board = [[] for _ in range(n+1)]
        for _ in range(k):
            x, y = map(int, input().split())
            board[y].append(x)

        w = int(input())
        q = deque([(w, 0)])
        answer = []
        while q:
            now, time = q.popleft()
            time += t[now]
            if len(board[now]) == 0 :
                answer.append(time)
            else:
                for to in board[now]:
                    q.append((to, time))

        answer.sort(reverse=True)
        print(answer[0])

solution()

File: baekjoon/java/src/gold3/Q1005_AcmCraft.java
package gold3;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// https://www.acmicpc.net/problem/1005
public class Q1005_AcmCraft {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int testCnt = Integer.parseInt(br.readLine());

        for (int i = 0; i < testCnt; i++) {
            List<Integer> time = new ArrayList<>();
            time.add(0);
            String[] meta = br.readLine().split(" ");
            int n = Integer.parseInt(meta[0]);
            int k = Integer.parseInt(meta[1]);
            String[] timeStrings = br.readLine().split(" ");
            for (String timeString : timeStrings) {
                time.add(Integer.parseInt(timeString));
            }

            List<List<Integer>> board = new ArrayList<>();
            int[] indegree = new int[n+1];
            for (int j = 0; j <= n; j++) {
                board.add(new ArrayList<>());
            }
            for (int j = 0; j < k; j++) {
                String[] order = br.readLine().split(" ");
                int later = Integer.parseInt(order[1]);
                board.get(Integer.parseInt(order[0])).add(later);
                indegree[later]++;
            }

            Integer[] result = time.toArray(new Integer[n+1]);
            int w = Integer.parseInt(br.readLine());
            Queue<Integer> queue = new LinkedList<>();
            for (int j = 1; j <= n; j++) {
                if(indegree[j] == 0) {
                    queue.add(j);
                }
            }
            while(!queue.isEmpty()) {
                int now = queue.poll();
                for (int to : board.get(now)) {
                    result[to] = Math.max(result[to], result[now] + time.get(to));
                    indegree[to]--;
                    if(indegree[to] == 0) {
                        queue.add(to);
                    }
                }
            }

            System.out.println(result[w]);
        }
    }
}
